bridges/snowbridge/pallets/system-v2/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+	v2::{Command, Message, SendMessage},
+	SendError,
+};
+use sp_core::H256;
+use sp_runtime::traits::MaybeEquivalence;
+use sp_std::prelude::*;
+use xcm::prelude::*;
+use xcm_executor::traits::ConvertLocation;
+
+#[cfg(feature = "runtime-benchmarks")]
+use frame_support::traits::OriginTrait;
+
+pub use pallet::*;
+
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+pub fn agent_id_of<T: Config>(location: &Location) -> Result<H256, DispatchError> {
+	T::AgentIdOf::convert_location(location).ok_or(Error::<T>::LocationConversionFailed.into())
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::dispatch::RawOrigin;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Send messages to Ethereum
+		type OutboundQueue: SendMessage;
+
+		/// Origin check for XCM locations that can create agents
+		type SiblingOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Converts Location to AgentId
+		type AgentIdOf: ConvertLocation<AgentId>;
+
+		type WeightInfo: WeightInfo;
+
+		/// This chain's Universal Location.
+		type UniversalLocation: Get<InteriorLocation>;
+
+		// The bridges configured Ethereum location
+		type EthereumLocation: Get<Location>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An CreateAgent message was sent to the Gateway
+		CreateAgent { location: Box<Location>, agent_id: AgentId },
+		/// Register Polkadot-native token as a wrapped ERC20 token on Ethereum
+		RegisterToken {
+			/// Location of Polkadot-native token
+			location: VersionedLocation,
+			/// ID of Polkadot-native token on Ethereum
+			foreign_token_id: H256,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		LocationConversionFailed,
+		AgentAlreadyCreated,
+		NoAgent,
+		UnsupportedLocationVersion,
+		InvalidLocation,
+		Send(SendError),
+	}
+
+	/// The set of registered agents
+	#[pallet::storage]
+	#[pallet::getter(fn agents)]
+	pub type Agents<T: Config> = StorageMap<_, Twox64Concat, AgentId, (), OptionQuery>;
+
+	/// Lookup table for foreign token ID to native location relative to ethereum
+	#[pallet::storage]
+	pub type ForeignToNativeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, TokenId, xcm::v5::Location, OptionQuery>;
+
+	/// Lookup table for native location relative to ethereum to foreign token ID
+	#[pallet::storage]
+	pub type NativeToForeignId<T: Config> =
+		StorageMap<_, Blake2_128Concat, xcm::v5::Location, TokenId, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
```

Comment:
@yrong also need to include `upgrade` and `set_operating_mode` extrinsics.

Reply:
Yeah, will address that in a seperated PR.